modules/replicator/src/main/java/org/apache/ignite/internal/replicator/ReplicaManager.java

```diff
@@ -1028,4 +1087,357 @@ public Set<ReplicationGroupId> startedGroups() {
+
+                if (context != null) {
+                    synchronized (context) {
+                        context.reservedForPrimary = false;
+                    }
+                }
+            }
+
+            return falseCompletedFuture();
+        }
+
+        ReplicaLifecycleContext getContext(ReplicationGroupId groupId) {
+            return replicaContexts.computeIfAbsent(groupId,
+                    // Treat the absence in the map as STOPPED.
+                    k -> new ReplicaLifecycleContext(ReplicaState.STOPPED, nullCompletedFuture()));
+        }
+
+        /**
+         * Can possibly start replica if it's not running or is stopping.
+         *
+         * @param groupId Group id.
+         * @param startOperation Replica start operation.
+         * @param forcedAssignments Assignments to reset forcibly, if needed. Assignments reset is only available when replica is started.
+         */
+        CompletableFuture<Boolean> weakReplicaStart(
+                ReplicationGroupId groupId,
+                Supplier<CompletableFuture<Boolean>> startOperation,
+                @Nullable Assignments forcedAssignments
+        ) {
+            ReplicaLifecycleContext context = getContext(groupId);
+
+            synchronized (context) {
+                ReplicaState state = context.replicaState;
+
+                log.debug("Weak replica start [grp={}, state={}, future={}].", groupId, state, context.previousOperationFuture);
+
+                if (state == ReplicaState.STOPPED || state == ReplicaState.STOPPING) {
+                    return startReplica(groupId, context, startOperation);
+                } else if (state == ReplicaState.ASSIGNED) {
+                    if (forcedAssignments != null) {
+                        assert forcedAssignments.force() :
+                                format("Unexpected assignments to force [assignments={}, groupId={}].", forcedAssignments, groupId);
+
+                        replicaManager.resetPeers(groupId, fromAssignments(forcedAssignments.nodes()));
+                    }
+
+                    // Telling the caller that the replica is started.
+                    return trueCompletedFuture();
+                } else if (state == ReplicaState.PRIMARY_ONLY) {
+                    context.replicaState = ReplicaState.ASSIGNED;
+
+                    return trueCompletedFuture();
+                } // else no-op.
+
+                throw new AssertionError("Replica start cannot begin while the replica is being started [groupId=" + groupId + "].");
+            }
+        }
+
+        private CompletableFuture<Boolean> startReplica(
+                ReplicationGroupId groupId,
+                ReplicaLifecycleContext context,
+                Supplier<CompletableFuture<Boolean>> startOperation
+        ) {
+            context.replicaState = ReplicaState.STARTING;
+            context.previousOperationFuture = context.previousOperationFuture
```

Comment:
I also believe that instead of exclusive states ASSIGNED and PRIMARY_ONLY that should be STARTED, STARTED_RESERVED, STARTED_PRIMARY or we may hide RESERVED/PRIMARY inside started.